### PR TITLE
WP 5.2 - 5.8: backport select changes to the test suite 

### PR DESCRIPTION
**Pull request by jrfnl**
Implementation of the backport proposal for WP 5.2 - 5.8 and lower.

### Build/Test Tools: Add Composer dependency on the PHPUnit Polyfills package.

The PHPUnit Polyfills package is an add-on for PHPUnit, which works around common issues for writing PHPUnit cross-version compatible tests.

Features:
* It offers a full set of polyfills for assertions and expectations introduced in PHPUnit since PHPUnit 4.8.
* It offers two generic TestCases which include these polyfills, but also solve the `void` return type issue for the fixtures methods.
* It offers a PHPUnit cross-version solution for the changes to the PHPUnit `TestListener` implementation.
* Supports PHPUnit 4.8 – current.
* Supports and is compatible with PHP 5.4 – current.

The package has no outside dependencies, other than PHPUnit, is actively maintained and endorsed by the maintainer of PHPUnit itself (the only package of its kind which has ever been endorsed).

While use of the package will not be actively implemented in WP < 5.9, adding the dependency will allow for future backports of (security) tests to WP 5.2 - 5.8 to be executed without the need for adjusting the tests.

The minimum supported PHPUnit version, as checked in the test bootstrap file, has been adjusted to comply with the minimum version supported by the PHPUnit Polyfills.

### Build: always run composer install for the test workflows

As the PHPUnit Polyfills is now a requirement, the test workflows need to run some form of `composer install` to make them available.
The tests themselves can, and will, still be run via a PHPUnit PHAR file.

Note: The version juggling done to get Composer to install on all PHP versions, means that in select cases, the `composer.lock` file and sometimes even the `composer.json` file needs to be changed on the fly.
This would then fail then build on the "Ensure version-controlled files are not modified" step.
To mitigate this, a very select `git checkout -- file...` command is added right after the Composer install to reset those files to their committed state.
This way any _real_ changes to version controlled files will still be picked up on, but the build won't fail on these changes made specifically for the Composer install.

Second note: alternatively, a choice could be made to forego the `composer install` in favour of cloning the Polyfills repo in a `vendor/yoast/phpunit-polyfills` directory for CI.
That should also allow for the tests to be able to run and would prevent making changes to the `composer.json`/`composer.lock` files.

### Tests: make the polyfills available to all tests

... and add loading of the polyfills to the test bootstrap with the same messages as per PR 1587.

### Tests: introduce wrapper method for the PHPUnit fixture methods

The wrapper methods are intended to make it easier for plugin/theme integration tests to make their test suite compatible with the changes in WordPress 5.9, while still allowing the test suite to be run on WP < 5.2.

Plugin/theme integration test suites which still want to run their tests against WP < 5.2 will need to handle conditionally making the assertion and expectation polyfill traits available based on the WP version on which the tests are being run from within their own test suite, but with these wrapper methods in place, the fixture method rename will no longer be problematic.

Notes:
As these wrapper methods use `static::...`, these wrapper methods will automatically call the camelCase method name in the "highest" class in the hierarchy, effectively making the wrapper methods available to all test case classes.

Trac ticket: https://core.trac.wordpress.org/ticket/53911

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/includes/bootstrap.php
Comment (desrosj) [inline tests/phpunit/includes/bootstrap.php:49]: Curious for the reasoning behind the patch version requirement for PHPUnit here. Will earlier versions of 5.7 not work? Is it because of the forward compatibility for `PHPUnit\Framework\(AssertionFailedError|Test|TestSuite)`?

**Review by desrosj — APPROVED** [review]
This seems good to me! Thanks for working on this.

File: tests/phpunit/includes/bootstrap.php
Comment (jrfnl) [reply tests/phpunit/includes/bootstrap.php:49]: Correct. Earlier versions won't work. PHPUnit added some forward-compatibility namespaced classes to the later patch versions in PHPUnit 4.8.x and 5.7.x and there are differences in which namespaced classes are available between the patch versions.
PHPUnit 5.7.21 is the first 5.x version which has enough of the namespaced classes available for the complete set of polyfills to work.

Comment (jrfnl) [reply tests/phpunit/includes/bootstrap.php:49]: Generally speaking, with a PHPUnit version as old as that, I would expect people to use the last version anyway, which is 5.7.27.

**Comment by hellofromtonya** [comment]
Waiting for PR #1587 to be merged before doing a final review of this PR. Why? Changes in that PR will affect this one.

**Comment by jrfnl** [comment]
@hellofromtonya FYI: the bootstrap messaging in this PR is already aligned with the bootstrap changes proposed in #1587, but yes, if #1587 would require further changes, this PR needs an update too.

**Comment by jrfnl** [comment]
FYI: Rebased + updated the "Tests: make the polyfills available to all tests" commit with the additional changes which were made for #1587 as those have now been committed.

**Review by hellofromtonya — APPROVED** [review]
Alrighty this PR now has all of the goodies in it. Looks ready for backporting 👍

**Review by swissspidy — APPROVED** [review]

**Review by hellofromtonya — APPROVED** [review]
Approving latest change: refined test bootstrap `composer install`  comments and command-line error message.

**Comment by hellofromtonya** [comment]
Committed via changesets [51838-51840,51843-51848] https://core.trac.wordpress.org/log/?revs=51838-51840,51843-51848

**Pull request closed without merging**
